File: modelo.java
/*
Ciencia da Computação - Pucminas
AEDs II
Pedro Henrique Lima Carvalho
Matricula: 651230
TP 3 - Q04
*/

import java.util.Date;


/**
* Classe Personagem
*/
class Personagem {

//atributos
	private String nome;
	private int altura;
	private double peso;
	private String corDoCabelo;
	private String corDaPele;
	private String corDosOlhos;
	private String anoNascimento;
	private String genero;
	private String homeworld;

//construtores
	Personagem(String nome, int altura, double peso, String corDoCabelo,
		   String corDaPele, String corDosOlhos, String anoNascimento,
		   String genero, String homeworld){

		this.nome = nome;
		this.altura = altura;
		this.peso = peso;
		this.corDoCabelo = corDoCabelo;
		this.corDaPele = corDaPele;
		this.corDosOlhos = corDosOlhos;
		this.anoNascimento = anoNascimento;
		this.genero = genero;
		this.homeworld = homeworld;
	}

	Personagem(String endereco){
		Arq.openRead(endereco);
		String lida = Arq.readLine();
		Arq.close();
		
		String[] dados = parsePersonagem(lida);
	
		this.nome = dados[0];
		try{
			this.altura = Integer.parseInt(dados[1]);
		}
		catch(Exception e){
			this.altura = 0;
		}
		try{
			dados[2]=dados[2].replaceAll(",", "");
			this.peso = Double.parseDouble(dados[2]);
		}
		catch(Exception e){
			this.peso = 0;
		}
		this.corDoCabelo = dados[3];
		this.corDaPele = dados[4];
		this.corDosOlhos = dados[5];
		this.anoNascimento = dados[6];
		this.genero = dados[7];
		this.homeworld = dados[8];		
	}

//metodos
 //gets
	/*
	*getNome
	*@return String nome
	*/
	public String getNome(){
		return(this.nome);
	}

	/*
	*getAltura
	*@return int altura
	*/
	public int getAltura(){
		return(this.altura);
	}

	/*
	*getPeso
	*@return double peso
	*/
	public double getPeso(){
		return(this.peso);
	}
	
	/*
	*get corDoCabelo
	*@return String corDoCabelo 
	*/
	public String getCorDoCabelo(){
		return(this.corDoCabelo);
	}

	/*
	*getCorDaPele
	*@return String corDaPele
	*/
	public String getCorDaPele(){
		return(this.corDaPele);
	}

	/*
	*getCorDosOlhos
	*@return String corDosOlho
	*/
	public String getCorDosOlhos(){
		return(this.corDosOlhos);
	}

	/*
	*getAnoNascimento
	*@return String anoNascimento
	*/
	public String getAnoNascimento(){
		return(this.anoNascimento);
	}

	/*
	*getGenero
	*@return String genero
	*/
	public String getGenero(){
		return(this.genero);
	}

	/*
	*getHomeworld
	*@return String homeworld
	*/
	public String getHomeworld(){
		return(this.homeworld);
	}

//sets
	/*
	*setNome
	*@param String nome
	*/
	public void setNome(String nome){
		this.nome = nome;
	}	

	/*
	*setAltura
	*@param int altura
	*/
	public void setAlgura(int altura){
		this.altura = altura;
	}

	/*
	*setPeso
	*@param double peso
	*/
	public void setPeso(double peso){
		this.peso = peso;
	}

	/*
	*setCorDoCabelo
	*@param String corDoCabelo
	*/
	public void setCorDoCabelo (String corDoCabelo){
		this.corDoCabelo = corDoCabelo;
	}

	/*
	*setCorDaPele
	*@param String corDaPele
	*/
	public void setCorDaPelo(String corDaPelo){
		this.corDaPele = corDaPele;
	}

	/*
	*setCorDosOlhos
	*@param String corDosOlhos
	*/
	public void setCorDosOlhos(String corDosOlhos){
		this.corDosOlhos = corDosOlhos;
	}

	/*
	*setAnoNascimento
	*@param String anoNascimento
	*/
	public void setAnoNascimento(String anoNascimento){
		this.anoNascimento = anoNascimento;
	}

	/*
	*setGenero
	*@param String genero
	*/
	public void setGenero(String genero){
		this.genero = genero;
	}

	/*
	*setHomeworld
	*@param String homeworld
	*/
	public void setHomeworld(String homeworld){
		this.homeworld = homeworld;
	}

//clone
	/*
	*clone
	*@param Personagem
	*@return Personagem
	*/
	public Personagem clone(Personagem a){
		return (new Personagem(a.nome, a.altura, a.peso, a.corDoCabelo,
			a.corDaPele, a.corDosOlhos, a.anoNascimento, a.genero,
			a.homeworld));
	}

//imprimir
	/*
	*imprimir
	*/
	public void imprimir(){
		MyIO.print(" ## "+this.nome);
		MyIO.print(" ## "+this.altura);
		if (this.peso%1==0)
			MyIO.print(" ## "+(int)this.peso);
		else
			MyIO.print(" ## "+this.peso);
		MyIO.print(" ## "+this.corDoCabelo);
		MyIO.print(" ## "+this.corDaPele);
		MyIO.print(" ## "+this.corDosOlhos);
		MyIO.print(" ## "+this.anoNascimento);
		MyIO.print(" ## "+this.genero);
		MyIO.print(" ## "+this.homeworld);
		MyIO.println(" ## ");

	}

//ler
	/*
	*ler
	*@return String com dados do personagem
	*/
	public String ler(){
		String s = " ## "+this.nome+" ## "+this.altura+" ## ";

		if (this.peso%1==0)
			s = s + (int)this.peso;
		else
			s = s + this.peso;
		s = s + " ## "+this.corDoCabelo+" ## "+this.corDaPele+" ## "+ 
		        this.corDosOlhos+" ## "+this.anoNascimento+" ## "+this.genero+
		        " ## "+this.homeworld+" ## ";	
		return s;
	}

//compareTo
	/*
	*@param Personagem a ser comparado e int da operacao
	*@return <0 se this menor q param, 0 se igual, >0 se maior
	*/
	public double compareTo (Personagem p, int op){
		double resp = 0.0;
	
		switch (op){	
			case 1:
  			 resp = this.altura - p.altura;
			 break;
			case 2:
			 resp = this.peso - p.peso;
			 break;
			case 3:
			 resp = this.corDoCabelo.compareTo(p.corDoCabelo);
			 break;
			case 4:
			 resp = this.corDaPele.compareTo(p.corDaPele);
			 break;
			case 5:
			 resp = this.corDosOlhos.compareTo(p.corDosOlhos);
			 break;
			case 6:
			 resp = this.anoNascimento.compareTo(p.anoNascimento);
			 break;
			case 7:
			 resp = this.genero.compareTo(p.genero);
			 break;
			case 8:
			 resp = this.homeworld.compareTo(p.homeworld);
			 break;
			default:
				resp = this.nome.compareTo(p.nome);
		}
		
		if (resp==0){
			resp = this.nome.compareTo(p.nome);
		}
		return resp;
	}	

//metodos estaticos publicos

 //isFim
	/**
	*isFim - verifica FIM
	*@param String
	*@return boolean
	*/
	public static boolean isFim(String s){
		return (s.equals("FIM"));
	}

 //toIso
	/**
	*toIso - muda o encoding de uma String para ISO
	*@param String UTF-8
	*@return String ISO
	*/
	public static String toIso(String s) throws Exception{
		return (new String(s.getBytes("UTF-8"), "ISO-8859-1"));
	}

 //toUtf
	/**
	*toUtf - muda o encoding de uma String para UTF-8
	*@param String ISO
	*@return String UTF-8
	*/
	public static String toUtf(String s) throws Exception{
		return(new String(s.getBytes("ISO-8859-1"), "UTF-8"));
	}

 //parsePersonagem
	/**
	*parsePersonagem - Separa os atributos do personagem de uma String
	*@param String
	*@return String[]
	*/
	private static String[] parsePersonagem(String s){
		String[] dados = new String[9];

		int init = 0;
		int end = -1;
		
		for (int i=0; i<9; i++){
			init = s.indexOf("\'", end+1)+1;
			end = s.indexOf("\'", init);
			init = s.indexOf("\'", end+1)+1;
			end = s.indexOf("\'", init);
			dados[i] = s.substring(init, end);
		}
		return (dados);
	}
}

/**
*Classe Celula
*/
class Celula{
 //atributos
	public Celula ant;
	public Personagem elemento;
	public Celula prox;

 //construtores
	public Celula(Personagem p){
		this.ant=null;
		this.elemento=p;
		this.prox=null;
	}
	
	public Celula(){
		this.ant=null;
		this.elemento=null;
		this.prox=null;
	}
}

/**
*Classe Lista Flexivel
*/
class Lista{
 //atributos
	private Celula primeiro;
	private Celula ultimo;

 //contrutotres
	public Lista(){
		this.primeiro = new Celula();
		this.ultimo = primeiro;
	}

//inserir
	/**
	*inserirInicio - insere um elemento no inicio da lista
	*@param Personagem elemento
	*/
	public void inserirInicio(Personagem p){
		Celula tmp = new Celula(p);
		tmp.prox=primeiro.prox;
		tmp.ant=primeiro;
		primeiro.prox=tmp;
		if (primeiro==ultimo)
			ultimo=primeiro.prox;
		else
			tmp.prox.ant=tmp;
		tmp=null;
	}

	/**
	*inserirFim - insere um elemento no fim da lista
	*@param Personagem elemento
	*/
	public void inserirFim(Personagem p){
		ultimo.prox = new Celula(p);
		ultimo.prox.ant=ultimo;
		ultimo=ultimo.prox;

	}

	/**
	*tamanho - retorna o tamanho da lista
	*@return tamanho int
	*/
	private int tamanho(){
		int resp = 0;
		for(Celula i=primeiro.prox; i!=null; i=i.prox){
			resp++;
		}
		return resp;
	}

	/**
	*inserir - insere um elemento em determinada posicao
	*@param Personagem elemento, int pos;
	*/
	public void inserir(Personagem p, int pos) throws Exception{
		int tamanho = tamanho();
		
		if (pos>tamanho || pos<0)
			throw new Exception ("Posicao Inexistente");
		else if (pos==tamanho)
			inserirFim(p);
		else if (pos==0)
			inserirInicio(p);
		else{
			Celula i=primeiro;
			for(int j=0; j<pos; j++, i=i.prox);
			
			Celula tmp = new Celula(p);
			tmp.prox = i.prox;
			tmp.ant=i;
			tmp.prox.ant=tmp;
			i.prox=tmp;
			i=tmp=null;
		}

	}

 //remover
	/**
	*removerInicio - remove um elemento do inicio da lista
	*@return Personagem elemento
	*/
	public Personagem removerInicio() throws Exception{
		if(primeiro==ultimo)
			throw new Exception("Lista Vazia");
		
		Celula tmp = primeiro;
		primeiro = primeiro.prox;
		tmp.prox=null;
		primeiro.ant=null;
		tmp=null;
		return (primeiro.elemento);
		
	}

	/*removerFim - remove um elemento do fim da lista
	*@return Personagem elemento
	*/
	public Personagem removerFim() throws Exception{
		if(primeiro==ultimo)
			throw new Exception("Lista Vazia");
		
		Personagem p=ultimo.elemento;
	
		ultimo=ultimo.ant;
		ultimo.prox.ant=null;
		ultimo.prox=null;		
	
		return p;
	}
	
	/*remover - remove um elemento da posicao
	*@param int posicao
	*@return Personagem p
	*/
	public Personagem remover(int pos) throws Exception{
		int tamanho = tamanho();
		Personagem p;

		if(pos<0 || pos>tamanho-1)
			throw new Exception("Posicao Inexistente");
		else if (pos==0)
			p=removerInicio();
		else if (pos==tamanho-1)
			p=removerFim();
		else{
			Celula i = primeiro.prox;
			for(int j=0; j<pos; j++, i=i.prox);
			
			p=i.elemento;

			i.prox.ant=i.ant;
			i.ant.prox=i.prox;
				
			i.ant=i.prox=null;
			i=null;
		}
		
		return p;
	}

 //mostrar
	/**
	*mostrar lista
	*/
	public void mostrar(){
		int j=0;
		for (Celula i=primeiro.prox; i!=null; i=i.prox, j++){
			MyIO.println("["+j+"] "+ i.elemento.ler());
		}
	}

	/**
	*mostrar lista sem posicao
	*/
	public void mostrar(char c){
		for(Celula i=primeiro.prox; i!=null; i=i.prox){
			MyIO.println(i.elemento.ler());
		}
	}

 //comandos
	/**
	*comandos - executa os comandos contidos em um string
	*@param String com comandos
	*/
	public void comandos(String s) throws Exception{
		String[] parsed = s.split(" ");
		
		if (parsed[0].equals("II")){
			this.inserirInicio(new Personagem(parsed[1]));
		}
		else if (parsed[0].equals("IF")){
			this.inserirFim(new Personagem(parsed[1]));
		}
		else if (parsed[0].equals("I*")){
			this.inserir(new Personagem(parsed[2]), Integer.parseInt(parsed[1]));
		}
		else if (parsed[0].equals("RI")){
			Personagem p = this.removerInicio();
			MyIO.println("(R) "+p.getNome());
		}
		else if (parsed[0].equals("RF")){
			Personagem p = this.removerFim();
			MyIO.println("(R) "+p.getNome());
		}		
		else if (parsed[0].equals("R*")){
			Personagem p = this.remover(Integer.parseInt(parsed[1]));
			MyIO.println("(R) "+p.getNome());

		}
	}
	
 //zerarPeso
	/**
	*zerarPeso - zera o peso dos elementos da lista
	*/
	public void zerarPeso(){
		for (Celula i=primeiro.prox; i!=null; i=i.prox){
			i.elemento.setPeso(0.0);
		}
	}
 
 //ordernar
	/**
	*swap - troca deois elementos da lista de posicao
	*@param Celula i, Celula j
	*/
	public void swap(Celula i, Celula j){
		Personagem buffer = i.elemento;
		i.elemento=j.elemento;
		j.elemento=buffer;
	}
	

	/**
	*ordernarQuickSort - ordena a lista pelo metodo QuickSort
	*@param int[] log para estatistica
	*/
	public void ordenarQuickSort(int[] log) throws Exception {
		if (primeiro==ultimo)
			throw new Exception ("Lista Vazia");
		quickSort(primeiro.prox, ultimo, log);
	}

	/**
	*goTo - retorna o ponteiro da celula entre outras duas
	*@param Celula i, Celula j
	*@return Celula meio;
	*/
	public Celula goTo (Celula i, Celula j){
		int tam = 0;
		Celula tmp;

		for (tmp=i; tmp!=j; tmp=tmp.prox, tam++);
		tmp=i;
		tam = tam/2;

		for (int k=0; k<tam; k++, tmp=tmp.prox);

		return tmp;
	}


	/**
	*quickSort - quickSort para lista flexivel duplamente encadeada
	*@param Celula esq, Celula dir, int[] log
	*/
	public void quickSort(Celula esq, Celula dir, int[] log){
		Personagem pivo = goTo(esq, dir).elemento;
		Celula i = esq;
		Celula j = dir;

		while (j.prox!=i && i.ant!=j && j.prox!=i.ant){
	
			log[0]+=2;
			while(i.elemento.compareTo(pivo, 3)<0){
				i=i.prox;
				log[0]++;
			}
			while(j.elemento.compareTo(pivo, 3)>0){
				j=j.ant;
				log[0]++;
			}
			
			if(j.prox!=i && i.ant!=j && j.prox!=i.ant){
				swap(i, j);
				i=i.prox;
				j=j.ant;
				log[1]+=3;
			}
		}

		if(i!=dir && i!=null){
			quickSort(i, dir, log);
		}
		if(j!=esq && j.prox!=esq){
			quickSort(esq, j, log);
		}
	}

}


/**
*Classe Main
*/
public class TP03Q07{
  		
	/**
	*tempo
	*@return long tempo
	*/
	public static long tempo(){
		return(new Date().getTime());
	}

				
	/**
	*Metodo main
	*/
	public static void main(String[] args) throws Exception{
		Lista list = new Lista();
		int[] log = new int[2];
		long inicio, fim;
		
		String input = MyIO.readLine();
		input = Personagem.toUtf(input);
		while(!Personagem.isFim(input)){
			list.inserirFim( new Personagem(input));
			input = MyIO.readLine();
			input = Personagem.toUtf(input);
		}
/*		
		int n = MyIO.readInt();
		
		for (int i=0; i<n; i++){
			input = MyIO.readLine();
			input = Personagem.toUtf(input);
			list.comandos(input);	
		}
*/
		inicio=tempo();
		list.ordenarQuickSort(log);
		fim=tempo();
		double segundos = ((double)(fim-inicio))/1000.0;
		
		//arquivo log
		Arq.openWrite("651230_quicksort3.txt");
		Arq.print("651230\t"+log[0]+"\t"+log[1]+"\t"+segundos);
		Arq.close();		

		list.mostrar('c');
	}	


}

File: modelo.cpp
/*
Ciencia da Computacao - Pucminas
Pedro Henrique Lima Carvalho
Matricula: 651230
AEDs 2

TP03 - Q04
*/

//dependencias
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// ----------------------------------- Personagem  ---------------------------------

/**
* Struct Personagem
*/
typedef struct Personagem{

//atributos
	char* nome;
	int altura;
	double peso;
	char* corDoCabelo;
	char* corDaPele;
	char* corDosOlhos;
	char* anoNascimento;
	char* genero;
	char* homeworld;

} Personagem;


//declaracoes

void mallocPerson (Personagem* p_person);

void preencherPersonagem(Personagem* p_person, char* s);


//construtores

/*
*constructor1
*@param todos atributos
*@return Personagem*
*/
Personagem* constructor1 (char* nome, int altura, double peso, char* corDoCabelo,
		   char* corDaPele, char* corDosOlhos, char* anoNascimento,
		   char* genero, char* homeworld){

	Personagem* p_person = (Personagem*) malloc(sizeof(Personagem)*1);
	p_person->nome = nome;
	p_person->altura = altura;
	p_person->peso = peso;
	p_person->corDoCabelo = corDoCabelo;
	p_person->corDaPele = corDaPele;
	p_person->corDosOlhos = corDosOlhos;
	p_person->anoNascimento = anoNascimento;
	p_person->genero = genero;
	p_person->homeworld = homeworld;

	return(p_person);
}

/*
*constructor2
*@param char* endereco arquivo
*@return Personagem
*/
Personagem* constructor2(char* endereco){
	//declaracoes
	FILE* arquivo = fopen(endereco, "r");
	Personagem* p_person = (Personagem*) malloc(sizeof(Personagem)*1);

	//teste
	if (arquivo==NULL)
		printf("%s\n", "Erro ao abrir arquivo");
	else{
		char* lida = (char*) malloc(sizeof(char)*1000);
		fgets(lida, 999, arquivo); 	
		lida[strlen(lida)-1]='\0';
		
		//alocar memoria
		mallocPerson(p_person);

		//preencher personagem			
		preencherPersonagem(p_person, lida);		
	}

	return (p_person);
}

/*
*removeChar
*@param char, char*
*/
void removeChar (char c, char* s){
	int pre = 0;
	int pos = 0;
		
	while (s[pre]!='\0'){
		if (s[pre]!=c){
			s[pos] = s[pre];
			pos++;
		}
		pre++;
	}
	s[pos]='\0';
}

/*
*mallocPerson
*@param Personagem*
*/
void mallocPerson (Personagem* p_person){
	p_person->nome = (char*) malloc(sizeof(char)*100);
	p_person->corDoCabelo = (char*) malloc(sizeof(char)*100);
	p_person->corDaPele = (char*) malloc(sizeof(char)*100);
	p_person->corDosOlhos = (char*) malloc(sizeof(char)*100);
	p_person->anoNascimento = (char*) malloc(sizeof(char)*100);
	p_person->genero = (char*) malloc(sizeof(char)*100);
	p_person->homeworld = (char*) malloc(sizeof(char)*100);
}

/*
*freePerson
*@param Personagem*
*/
void freePerson (Personagem* p_person){
	free(p_person->nome);
	free(p_person->corDoCabelo);
	free(p_person->corDaPele);
	free(p_person->corDosOlhos);
	free(p_person->anoNascimento);
	free(p_person->genero);
	free(p_person->homeworld);

}


/*
*clone
*@param Personagem*
*@return Personagem*
*/
Personagem* clone(Personagem* p_person){
	//delcaracoes
	Personagem* p_clone = (Personagem*) malloc(sizeof(Personagem)*1);
	
	//alocar memoria para clone
	mallocPerson(p_clone);
	
	strcpy(p_clone->nome, p_person->nome);
	p_clone->altura = p_person->altura;	
	p_clone->peso = p_person->peso;
	strcpy(p_clone->corDoCabelo, p_person->corDoCabelo);
	strcpy(p_clone->corDaPele, p_person->corDaPele);
	strcpy(p_clone->corDosOlhos, p_person->corDosOlhos);
	strcpy(p_clone->anoNascimento, p_person->anoNascimento);
	strcpy(p_clone->genero, p_person->genero);
	strcpy(p_clone->homeworld, p_person->homeworld);
	
	return (p_clone);
}

/*
*imprimir
*/
void imprimir(Personagem* p_person){

	printf("%s%s", " ## ", p_person->nome);
	printf("%s%d", " ## ", p_person->altura);
	printf("%s%g", " ## ", p_person->peso);
	printf("%s%s", " ## ", p_person->corDoCabelo);
	printf("%s%s", " ## ", p_person->corDaPele);
	printf("%s%s", " ## ", p_person->corDosOlhos);
	printf("%s%s", " ## ", p_person->anoNascimento);
	printf("%s%s", " ## ", p_person->genero);
	printf("%s%s", " ## ", p_person->homeworld);
	printf(" ## \n");

	}


/*
*ler
*@param Personagem*
*@return char*
*/
char*  ler(Personagem* p_person){
	char buffer[50];
	char* lida = (char*) malloc(sizeof(char)*1000);
	
	strcat(lida, " ## ");
	strcat(lida, p_person->nome);
	
	strcat(lida, " ## ");
	sprintf(buffer, "%d", p_person->altura);
	strcat(lida, buffer);	
	
	strcat(lida, " ## ");
	strcat(lida, buffer);
	sprintf(buffer, "%g", p_person->peso);
	
	strcat(lida, " ## ");
	strcat(lida, p_person->corDoCabelo);

	strcat(lida, " ## ");
	strcat(lida, p_person->corDaPele);
	
	strcat(lida, " ## ");
	strcat(lida, p_person->corDosOlhos);
	
	strcat(lida, " ## ");
	strcat(lida, p_person->anoNascimento);
	
	strcat(lida, " ## ");
	strcat(lida, p_person->genero);
	
	strcat(lida, " ## ");
	strcat(lida, p_person->homeworld);

	strcat(lida, " ## ");
			
	return(lida);	
}

/**
*isFim - verifica FIM
*@param char*
*@return boolean
*/
bool isFim(char* s){
	return (strcmp(s, "FIM") == 0);
}


/**
*preencherPersonagem - Separa os atributos do personagem de uma String
*@param Personagem*, char*
*/
void preencherPersonagem(Personagem* p_person, char* s){
	char buffer[9][100];
	int c_buffer= 0;
	int c_s=0;
	int init = 0;
	int end = -1;
		
	for (int i=0; i<9; i++){
		init=end+1;
		while (s[init]!='\''){
			init++;
		}
		end=init+1;
		while (s[end]!='\''){
			end++;
		}
		init=end+1;
		while (s[init]!='\''){
			init++;
		}
		end=init+1;
		while (s[end]!='\''){
			end++;
		}
		
		c_s=init+1;
		while(c_s<end){
			buffer[i][c_buffer] = s[c_s];
			buffer[i][c_buffer+1]='\0';
			c_buffer++;
			c_s++;
		}
		c_buffer=0;
	}

	strcpy(p_person->nome, buffer[0]);
	
	if (0 == sscanf(buffer[1], "%d", &p_person->altura))//testar unknown
		p_person->altura=0;

	removeChar(',', buffer[2]); //retirar virgula do peso	
	if (0 == sscanf(buffer[2], "%lf", &p_person->peso))//testar unknown
		p_person->peso=0;

	strcpy(p_person->corDoCabelo, buffer[3]);
	strcpy(p_person->corDaPele, buffer[4]);
	strcpy(p_person->corDosOlhos, buffer[5]);
	strcpy(p_person->anoNascimento, buffer[6]);
	strcpy(p_person->genero, buffer[7]);
	strcpy(p_person->homeworld, buffer[8]);
		
}

/**
*compare - compara dois Personagens de acordo com a operacao. Nome como desempate
*@param Lista*, Persongem* 1, Personagem* 2, int operacao
*@return double <0 se 1 menor que 2, 0 se iguais, >0 se 1 maior que 2
*/
double compare(Personagem* p_person1, Personagem* p_person2, int op){
	double resp = 0;
	
	switch(op){
		case 1:
		 resp = p_person1->altura - p_person2->altura;
		 break;
		case 2:
		 resp = p_person1->peso - p_person2->peso;
		 break;
		case 3:
		 resp = strcmp(p_person1->corDoCabelo, p_person2->corDoCabelo);
		 break;
		case 4:
		 resp = strcmp(p_person1->corDaPele, p_person2->corDaPele);
		 break;
		case 5:
		 resp = strcmp(p_person1->corDosOlhos, p_person2->corDosOlhos);
		 break;
		case 6:
		 resp = strcmp(p_person1->anoNascimento, p_person2->anoNascimento);
		 break;
		case 7:
		 resp = strcmp(p_person1->genero, p_person2->genero);
		 break;
		case 8:
		 resp = strcmp(p_person1->homeworld, p_person2->homeworld);
		 break;
		default:
		 resp = strcmp(p_person1->nome, p_person2->nome);

	}
	
	if (resp == 0){
		resp = strcmp(p_person1->nome, p_person2->nome);
	}

	return resp;
} 


//------------------------------------ Celula ----------------------------------------------

/**
*struct Celula
*/
typedef struct Celula{
 
 //atributos
	Celula* ant;
	Personagem* elemento;
	Celula* prox;

}Celula;

 //construtor
	Celula* construtorCelula (Personagem* p){
		Celula* cel = (Celula*) malloc(sizeof(Celula));
		
		cel->ant = NULL;
		cel->elemento = p;
		cel->prox = NULL;
		
		return cel;
	}

//------------------------------- Lista Flexivel -------------------------------------------

/**
*struct Lista
*/
typedef struct Lista{
 
//atributos
	Celula* primeiro;
	Celula* ultimo;
}Lista;

//construtor
/**
*construtorLista - inicializa uma Lista
*@param int tamanho
*@return list*
*/
Lista* construtorLista (){

	Lista* p_lista;
	p_lista = (Lista*) malloc (sizeof(Lista));	
	p_lista->primeiro = construtorCelula (NULL);
	p_lista->ultimo = p_lista->primeiro;
	
	return p_lista;
}

//metodos

/**
*removerInicio - remove um elemento inicio da lista
*@return Personagem*
*/
Personagem* RemoverInicio (Lista* p_lista){
	Personagem* p_person;

	if (p_lista->primeiro == p_lista->ultimo){
		printf("%s\n", "Erro - Lista Vazia");
	}
	else{
		Celula* tmp = p_lista->primeiro;
		p_lista->primeiro = p_lista->primeiro->prox;
		p_person = p_lista->primeiro->elemento;
		p_lista->primeiro->ant=NULL;
		p_lista->primeiro->elemento=NULL;
		free(tmp);
		
	}
	return p_person;
}


/**
*inserirFim - insere um elemento no final da lista
*@param Lista* Personagem*
*/
void inserirFim(Lista* p_lista, Personagem* p_person){
	
	p_lista->ultimo->prox = construtorCelula(p_person);
	p_lista->ultimo->prox->ant = p_lista->ultimo;
	p_lista->ultimo=p_lista->ultimo->prox;		
}

/**
*mostrar lista
*@param Lista*
*/
void mostrar(Lista* p_lista){
	int j=0;
	for (Celula* i=p_lista->primeiro->prox; i!=NULL; i=i->prox, j++){
		printf("%s%d%s", "[", j, "] ");
		imprimir(i->elemento);
	} 
}

/**mostrar2 - fila sem posicao
*@param Fila*
*/
void mostrar2(Lista* p_lista){
	for (Celula* i=p_lista->primeiro->prox; i!=NULL; i=i->prox){
		imprimir(i->elemento);
	}
}


/**
*freeLista - libera a memoria da Lista
*@param Lista*
*/
void freeLista (Lista* p_Lista){
	
	Celula* i = p_Lista->primeiro->prox;
	free(p_Lista->primeiro);
	p_Lista->ultimo=NULL;
	if (i!=NULL){
		for (Celula* j = i->prox; j!=NULL; j=j->prox){
			freePerson(i->elemento);
			free(i);
			i=j;
		}	
		freePerson(i->elemento);
		free(i);
	}
	free(p_Lista);
}

/**
*swap - troca de posicao dois elementos da lista
*@param Celula* i, Celula* j
*/
void swap(Celula* i, Celula* j){
	Personagem* buffer = i->elemento;
	i->elemento = j->elemento;
	j->elemento = buffer;
}

/**
*getPivo - retorna o Personagem* pivo
*@param Celula* i, Celula* j
*@return Personagem* p_person
*/
Personagem* getPivo (Celula* i, Celula* j){
	int tam = 0;
	Celula* tmp;
	for (tmp=i; tmp!=j; tmp=tmp->prox, tam++);

	tam=tam/2;
	
	tmp=i;
	for(int k=0; k<tam; k++){
		tmp=tmp->prox;
	}
	return(tmp->elemento);
}

/**
*QuickSort - quicksort
*@param Celula* esq, Celula* dir, int log[]
*/
void quickSort (Celula* esq, Celula* dir, int log[]){
	
	Celula* i = esq;
	Celula* j = dir;

	Personagem* pivo = getPivo(i, j);
       
	while (i->ant!=j && j->prox!=i && i->ant!=j->prox){
		
		log[0]+=2;		

		while (compare(i->elemento, pivo, 3)<0.0){
			i=i->prox;

			log[0]++;
		}
		while (compare(j->elemento, pivo, 3)>0.0){
			j=j->ant;

			log[0]++;
		}

		if(i->ant!=j && j->prox!=i && i->ant!=j->prox){
			swap(i,j);
			i=i->prox;
			j=j->ant;

			log[1]+=3;
		}
	}

        if (j->prox!=esq && j!=esq){
		quickSort(esq, j, log);
	}
	if (i!=dir && i!=NULL){
		quickSort(i, dir, log);
	}

	
}

/**
*ordenarQuickSort - ordena a lista pelo metodo quicksort
*@param Lista* p_lista, int log[]
*/	
void ordenarQuickSort(Lista* p_lista, int log[]){
	quickSort(p_lista->primeiro->prox, p_lista->ultimo, log);
}

//----------------------------------- Main ---------------------------------


/**
*Metodo main
*/
int  main(void){
        
	Lista* p_lista = construtorLista();
	int log[] = {0,0};
	Personagem* p;
	char* input = (char*) malloc(sizeof(char) * 100);

	fgets(input, 99, stdin);
	input[strlen(input)-1]='\0';
		
	while(!isFim(input)){
		p = constructor2(input);
		inserirFim(p_lista, p);
		fgets(input, 99, stdin);
		input[strlen(input)-1]='\0';
	}	
	clock_t inicio = clock();
	ordenarQuickSort(p_lista, log);
	clock_t fim = clock();

	mostrar2(p_lista);

	double segundos = (fim - inicio) / (double)CLOCKS_PER_SEC / 1000.0;

	//arquivo log
	FILE* arq = fopen("651230_quicksort2.txt", "wt");
	fprintf (arq, "%s\t%d\t%d\t%lf","651230", log[0], log[1], segundos);
	fclose(arq);

	freeLista(p_lista);
}	


